src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,192 @@
+        return message.foregroundRole();
+    case Qt::UserRole:
+        return message.userRole();
+    case TimeRole:
+        return message.time();
+    case MessageRole:
+        return message.message();
+    default:
+        return {};
+    }
+}
+
+void BaseLogModel::addNewMessage(const BaseLogModel::Message &message)
+{
+    // if row is inserted on filled up buffer, the size will not change but because of calling of beginInsertRows function
+    // we'll have ghost rows.
+    if (m_messages.size() == MAX_VISIBLE_MESSAGES) {
+        const int lastMessage = m_messages.size() - 1;
+        beginRemoveRows(QModelIndex(), lastMessage, lastMessage);
+        m_messages.pop_back();
+        endRemoveRows();
+    }
+
+    beginInsertRows(QModelIndex(), 0, 0);
+    m_messages.push_front(message);
+    endInsertRows();
+}
+
+void BaseLogModel::reset()
+{
+    beginResetModel();
+    m_messages.clear();
+    endResetModel();
+}
+
+LogMessageModel::LogMessageModel(QObject *parent)
+    : BaseLogModel(parent)
+{
+    for (const Log::Msg &msg : asConst(Logger::instance()->getMessages()))
+        handleNewMessage(msg);
+
+    // Logger::newLogMessage can be emitted from different threads, so to avoid need of synchronization we use Qt::QueuedConnection
+    connect(Logger::instance(), &Logger::newLogMessage, this, &LogMessageModel::handleNewMessage, Qt::QueuedConnection);
+}
+
+void LogMessageModel::handleNewMessage(const Log::Msg &log)
+{
+    const QString time = QDateTime::fromMSecsSinceEpoch(log.timestamp).toString(Qt::SystemLocaleShortDate);
+    const QString message = log.message;
+
+    QColor foreground;
+    switch (log.type) {
+    // The RGB QColor constructor is used for performance
+    case Log::NORMAL:
+        foreground = QApplication::palette().color(QPalette::WindowText);
+        break;
+    case Log::INFO:
+        foreground = QColor(0, 0, 255); // blue
+        break;
+    case Log::WARNING:
+        foreground = QColor(255, 165, 0);  // orange
+        break;
+    case Log::CRITICAL:
+        foreground = QColor(255, 0, 0);  // red
+        break;
+    default:
+        Q_ASSERT(false);
+        break;
+    }
+
+    addNewMessage({time, message, foreground, log.type});
+}
+
+LogPeerModel::LogPeerModel(QObject *parent)
+    : BaseLogModel(parent)
+{
+    for (const Log::Peer &peer : asConst(Logger::instance()->getPeers()))
+        handleNewMessage(peer);
+
+    // Logger::newLogPeer can be emitted from different threads, so to avoid need of synchronization we use Qt::QueuedConnection
+    connect(Logger::instance(), &Logger::newLogPeer, this, &LogPeerModel::handleNewMessage, Qt::QueuedConnection);
+}
+
+void LogPeerModel::handleNewMessage(const Log::Peer &peer)
+{
+    const QString time = QDateTime::fromMSecsSinceEpoch(peer.timestamp).toString(Qt::SystemLocaleShortDate);
+    const QString message = peer.blocked
+            ? tr("%1 was blocked %2", "0.0.0.0 was blocked due to reason").arg(peer.ip, peer.reason)
```

Comment:
Format should be either "%1 was blocked due to %2" or "%1 was blocked. Reason: %2".